service/history/mutableStateTaskRefresher.go

```diff
@@ -0,0 +1,394 @@
+		domainCache: domainCache,
+		eventsCache: eventsCache,
+		logger:      logger,
+
+		mutableState: mutableState,
+		taskGenerator: newMutableStateTaskGenerator(
+			domainCache,
+			logger,
+			mutableState,
+		),
+	}
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasks(nowTimestamp int64) error {
+
+	if err := r.refreshTasksForWorkflowStart(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForWorkflowClose(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForRecordWorkflowStarted(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForDecision(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForActivity(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForTimer(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForChildWorkflow(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForCancelExternalWorkflow(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForSignalExternalWorkflow(nowTimestamp); err != nil {
+		return err
+	}
+
+	if err := r.refreshTasksForWorkflowSearchAttr(nowTimestamp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForWorkflowStart(
+	nowTimestamp int64,
+) error {
+
+	startEvent, ok := r.mutableState.GetStartEvent()
+	if !ok {
+		return &shared.InternalServiceError{Message: "unable to load start event."}
+	}
+
+	if err := r.taskGenerator.generateWorkflowStartTasks(
+		nowTimestamp,
+		startEvent,
+	); err != nil {
+		return err
+	}
+
+	startAttr := startEvent.WorkflowExecutionStartedEventAttributes
+	if !r.mutableState.HasProcessedOrPendingDecision() && startAttr.GetFirstDecisionTaskBackoffSeconds() > 0 {
+		if err := r.taskGenerator.generateDelayedDecisionTasks(
+			nowTimestamp,
+			startEvent,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForWorkflowClose(
+	nowTimestamp int64,
+) error {
+
+	executionInfo := r.mutableState.GetExecutionInfo()
+
+	if executionInfo.CloseStatus != persistence.WorkflowCloseStatusNone {
+		return r.taskGenerator.generateWorkflowCloseTasks(
+			nowTimestamp,
+		)
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForRecordWorkflowStarted(
+	nowTimestamp int64,
+) error {
+
+	executionInfo := r.mutableState.GetExecutionInfo()
+
+	if executionInfo.CloseStatus == persistence.WorkflowCloseStatusNone {
+		return r.taskGenerator.generateRecordWorkflowStartedTasks(
+			nowTimestamp,
+		)
+	}
+
+	return nil
+}
+
+func (r *mutableStateTaskRefresherImpl) refreshTasksForDecision(
+	nowTimestamp int64,
+) error {
+
+	if !r.mutableState.HasPendingDecision() {
+		// no decision task at all
+		return nil
+	}
+
+	decision, ok := r.mutableState.GetPendingDecision()
+	if !ok {
+		return &shared.InternalServiceError{Message: "impossible case: cannot get pending decision"}
```

Comment:
it's better to remove "impossible case" because it doesn't give any valuable information. As an error, then it returns here, it is already possible.

Reply:
what if we actually have a bug? handling impossible case (i.e. return error rather than silently ignore the error) helps debugging, and in the future, refactoring.

Reply:
I understand what you are trying to give. This error is more critical than other errors like ConditionFailed/Timeout/ShardLost etc. When this error happens, it means we could have a bug. 

Imagine you run a command line and it returns an error telling you this is impossible... Whether it is impossible or not is too subjective to tell at that point. What you need is provide the fact what is causing error.

So I would suggest to reword it to "cannot get pending decision, it could be a bug". It will be more objective.

Reply:
done